src/wolff_producer.erl

```diff
@@ -116,40 +129,50 @@ stop(Pid) ->
 -spec send(pid(), [wolff:msg()], wolff:ack_fun()) -> ok.
 send(Pid, [_ | _] = Batch0, AckFun) ->
   Caller = self(),
-  Mref = erlang:monitor(process, Pid),
   Batch = ensure_ts(Batch0),
-  erlang:send(Pid, ?SEND_REQ({Caller, Mref}, Batch, AckFun)),
-  receive
-    {Mref, ?queued} ->
-      erlang:demonitor(Mref, [flush]),
-      ok;
-    {'DOWN', Mref, _, _, Reason} ->
-      erlang:error({producer_down, Reason})
+  case msg_batcher_object:get(Pid) of
+    not_found ->
+      Mref = erlang:monitor(process, Pid),
+      erlang:send(Pid, ?SEND_REQ({Caller, Mref}, Batch, AckFun)),
+      receive
+        {Mref, ?queued} ->
+          erlang:demonitor(Mref, [flush]),
+          ok;
+        {'DOWN', Mref, _, _, Reason} ->
+          erlang:error({producer_down, Reason})
+      end;
+    _ ->
+      case msg_batcher:enqueue(Pid, ?SEND_REQ(?no_queued_reply, Batch, AckFun)) of
+        ok -> ok;
+        {error, batcher_not_found} ->
+          erlang:throw(producer_down)
+      end
   end.
 
 %% @doc Send a batch synchronously.
 %% Raise error exception in case produce pid is down or when timed out.
 -spec send_sync(pid(), [wolff:msg()], timeout()) -> {partition(), offset_reply()}.
 send_sync(Pid, Batch0, Timeout) ->
   Caller = self(),
-  Mref = erlang:monitor(process, Pid),
+  ReqRef = erlang:make_ref(),
   %% synced local usage, safe to use anonymous fun
   AckFun = fun(Partition, BaseOffset) ->
-               _ = erlang:send(Caller, {Mref, Partition, BaseOffset}),
+               _ = erlang:send(Caller, {ReqRef, Partition, BaseOffset}),
                ok
            end,
   Batch = ensure_ts(Batch0),
-  erlang:send(Pid, ?SEND_REQ(?no_queued_reply, Batch, AckFun)),
-  receive
-    {Mref, Partition, BaseOffset} ->
-      erlang:demonitor(Mref, [flush]),
-      {Partition, BaseOffset};
-    {'DOWN', Mref, _, _, Reason} ->
-      erlang:error({producer_down, Reason})
-  after
-    Timeout ->
-      erlang:demonitor(Mref, [flush]),
-      erlang:error(timeout)
+  Request = ?SEND_REQ(?no_queued_reply, Batch, AckFun),
+  case msg_batcher_object:get(Pid) of
+    not_found ->
+      Mref = erlang:monitor(process, Pid),
+      erlang:send(Pid, Request),
+      monitored_wait_reply(Timeout, ReqRef, Mref);
```

Comment:
implementing the receive in another function will lose the `receive` compiler optimisation